### ISSUE TITLE
Adding Time column to already existing Table or QTable makes the column a Column object with dtype= 'object'

### ISSUE THREAD

**Issue by AustereCuriosity** [astropy/astropy#5911]
If I try to add a column to a Table using the following way, the column retains as a Time Object.
In [51]: t = Table()

In [52]: t['index'] = [1, 2]

In [53]: t['time'] = Time(['2001-01-02T12:34:56', '2001-02-03T00:01:02'])

In [54]: t
Out[54]: 
<Table length=2>
index                      time

int64                     object
-----           -----------------------

    1              2001-01-02T12:34:56.000
    2              `2001-02-03T00:01:02.000

In [55]: t['time']
Out[55]: <Time object: scale='utc' format='isot' value=['2001-01-02T12:34:56.000' '2001-02-03T00:01:02.000']>

However, if I try to add the column using add_column, for eg. if I want to add a column at a specified index, the column becomes a Column object with dtype= 'object'

In [41]: tab = Table()

In [42]: tab.add_column(Table.Column([1,2], unit='degree', name='wave'))

In [44]: tab.add_column(Table.Column(Time(['2001-01-02T12:34:56', '2001-02-03T00:01:02']), name='time'))

In [45]: tab
Out[45]: 
<Table length=2>
 wave           time

 deg 

int64          object
        
-----     -----------------------

    1        2001-01-02T12:34:56.000
    2        2001-02-03T00:01:02.000

In [68]: tab['time']
Out[68]: 
<Column name='time' dtype='object' length=2>
2001-01-02T12:34:56.000
2001-02-03T00:01:02.000

Thus if I try to write this table to a FITS file, instead of the "ValueError: cannot write table with mixin column(s) ['time']", it gives "ValueError: Illegal format object".

I came across this while understanding working with mixin columns, so pardon me if this isn't the way to deal with them. I couldn't find anything relevant in the docs.

**Comment by taldcroft** [comment]
@AustereCuriosity - the issue is that you converted the `Time` object into a `Column` prior to using `add_column`.  At that point it is just an ordinary column and not a mixin.  The `Column` class doesn't exactly know what to do with `Time` so it makes an `object` dtype array.  However, as long as the `Time` object has a name you can add it using `add_column`:
```
In [2]: t = simple_table()
In [3]: tm = Time([1,2,3], format='cxcsec')
In [4]: tm.info.name = 'd'
In [5]: t.add_column(tm)
In [6]: t
Out[6]: 
<Table length=3>
  a      b     c           d        
int64 float64 str1       object     
----- ------- ---- -----------------
    1     1.0    c 1.000000000001755
    2     2.0    d 1.999999999998714
    3     3.0    e 3.000000000000469

In [7]: type(t['d'])
Out[7]: astropy.time.core.Time
```

**Comment by mhvk** [comment]
This would be good addition to the documentation of `add_column`! (or should one just get a default name equal to the column number or so?)

**Comment by mhvk** [comment]
(I removed the `time` label since it is not particular to that class.)

**Comment by taldcroft** [comment]
@mhvk - yes, I like the idea of adding a default column name and making this work a bit more generally.  Or maybe just add a `name` kwarg?  I originally didn't want to slide down the slope of adding `Column` initializers (would we then want `unit` or `format`?), but in practice just the name would be enough.  We know this because the `t['new_col'] = new_col` is good enough 98% of the time except for putting the column in a particular location.

A more radical idea would be allowing a syntax like below and deprecating `add_column`:
```
t['new_col', 0] = new_col  # Add new_col at position 0
```
Not 100% sure this is a good idea in terms of readability, but just throwing it out there.

**Comment by mhvk** [comment]
@taldcroft - I think in almost all cases one does not care about the position of a column in a table, and in most of those where one does it would be order of insertion, which we have by default, so I think we should not go the `[name, place]` route. (Instead, one could make a new `move_column` or perhaps extend `rename_column` to allow one to give a new index.)

On adding the `name` argument: I think that is a better idea than my suggestion of generating a default, but the one thing I don't like is that `add_column` and `add_columns` then become asymmetric (Having a `name` argument for the latter would seem to add too much complexity). Since in the end, as you said, in most cases directly setting the column works, my instinct would be to leave that "the one obvious way to do it" and just add your "for mixin columns, set the name beforehand" instruction to both the `add_column(s)" docstring and the "modifying a table" documentation.

**Comment by taldcroft** [comment]
> I think in almost all cases one does not care about the position of a column in a table

For interactive work in a notebook and final presentation (writing out to latex), order can matter a fair bit (at least to me).  I use `add_column` with an explicit `index` every so often, mostly to get a column near the left side of the table where it gets more prominence, or to have particular columns next to each other.

About the asymmetry, note that `add_column` has `index` (scalar) and `add_columns` has `indexes` (which can be `None` or a list), so there is already precedent in the API for `name` and `names`.  Furthermore, `add_column` is actually just a call to `add_columns` with the scalar values passed in as a list, i.e. `add_columns([new_col], ..)`. So doing `name` for `add_column` almost requires adding `names` to `add_columns`.

For `names` the logic is pretty simple, namely check that it is either `None` or a sequence of the right length.  The only hitch I can think of is when renaming a column we probably are forced to make a copy.  I don't remember offhand if `add_columns` makes copies by default.  Anyway, that's a minor issue.

About `move_column`, that's a fine idea.

**Comment by AustereCuriosity** [comment]
Actually, my concern was the Table.Column() can understand Quantity objects and can convert them accordingly so shouldn't Time be recognised as well?

**Comment by taldcroft** [comment]
> Actually, my concern was that Table.Column() can understand Quantity objects and can convert them accordingly so shouldn't Time be recognised as well?

This is only possible because the meta associated with `Quantity` is a subset of possible `Column` meta.  This is not the case for `Time`, and that is one driver for the whole concept of mixin columns.

**Comment by AustereCuriosity** [comment]
@taldcroft Yes true. I was thinking instead of storing the entire metadata, which is definitely not possible, Column could be made intelligent to store a relevant unit that could atleast ensure that it is interpreted as a string containing time data. But I am not so sure how that could be done or is it even necessary.

I especially liked the idea of deprecating add_column and using the following:
t['new_col', 0] = new_col  # Add new_col at position 0

We can then include the idea of move_column(), so that if the user later decides that he wants to store an existing column at a different index, he can do so.

**Comment by taldcroft** [comment]
I think @mhvk convinced me that just extending the `add_column(s)` API would be better.  The non-standard `item` syntax I suggested was just a thought, but I am not sure I like it now.

**Comment by mhvk** [comment]
fixed by #5996
